docs(policy)_: Introduction of policy zero (#6165)
status-im/status-go: _docs/policies/submitting-policy.md

# Purpose

Policy Zero establishes the foundational guidelines for creating,
reviewing, and maintaining policies in the `status-go` Git repository.
This policy aims to create a collaborative, inclusive, and transparent
process for defining repository policies, specifically regarding how
developers engage with and contribute to the `status-go` repository.
For a more detailed description, please refer to [the README.md](./README.md).

# Submitting a Policy Proposal

- Any individual MAY propose a new policy.
- Policy ideas SHOULD be discussed with contributors who wish to have
a voice in how the repository operates, including Core Contributors
(CCs) and external contributors.
- All policies MUST be submitted to the `_docs/policies`
 directory as a pull request (PR) within the `status-go` repository.
- All policies MUST be in Markdown format.

> - Policy file names MUST follow [File name conventions for ADRs](https://github.com/joelparkerhenderson/architecture-decision-record?tab=readme-ov-file#file-name-conventions-for-adrs), e.g. `submitting-policy.md`.

Reviewer: Based on what's said on the link posted here, "submitting" is not an option, it should be "submit-policy.md", or we can change it to `propose-policy.md`, `introduce-policy.md` or something like that. Can we align on that naming?

Author: Hey @saledjenic that is fair and reasonable. I'll make this change.

Author: Done

- A policy MUST include a brief justification, addressing the question:
"Why has this policy been introduced?"

# Review and Approval Process

- Policy PRs SHOULD be reviewed by as many contributors as possible
who wish to engage in the process.
- Any CC MAY review, approve, and/or request changes to a policy
proposal PR.
- For any policy PR to be eligible for merging, it:
  - MUST address all feedback provided during the review process.

>   - MUST be approved by all team leads (of Status Desktop and Mobile).

Reviewer 2: ```suggestion
  - MUST be approved by all team leads (of Status Desktop,  Mobile and Waku Chat).
```

Author: Done

  - MUST be approved by all members of the @status-im/status-go-guild
  GitHub team.

>   - MUST receive a minimum of six approvals from CCs.

Reviewer 3: we mean this "in addition to the previous two points", right? Team Leads and Guild members are also CCs :D

Reviewer 4: AFAIK it is NOT in addition. It is intended to be resistant to changes in team leads and Status-go guild teams (which might decrease over time).

Reviewer 3: hmm interesting. I figured the intention was to make sure "common folk" are included on top of people belonging to special groups. We might've discussed this elsewhere already, I simply can't remember.

Reviewer 5: Yes, my understanding matches @osmaczko. 
We set the quorum to 6 in general. Yes, currently guilds members + team leads = 9 people, so the quorum will automatically gathered, but things might change in the future. And we show the intention of broad approvals, i.e. at least 6 members.

Reviewer 3: alright, even I might have agreed with this interpretation a couple of weeks ago. Forget my comment then!

Author: I think that there is enough risk of ambiguity here that I'll suggest something different:
```suggestion
  - MUST receive a minimum of six approvals, regardless of the number of team leads or members of the @status-im/status-go-guild GitHub team.
```

What do you think? cc @dlipicar @osmaczko @igor-sirotin

# Review and Approval Process

The core function of the review and approval process for policy
PRs is to reach consensus on any issue and to reflect the range
of perspectives among all contributors who wish to have a voice
in how the `status-go` repository operates, including both Core
Contributors (CCs) and external contributors. Policy submissions
must aim to achieve broad contributor support and give key stakeholders
a chance to gain context of the policy requirements.

- Policy PRs SHOULD be reviewed by as many contributors as possible
who wish to engage in the process.
- Any CC MAY review, approve, and/or request changes to a policy
proposal PR.
- For any policy PR to be eligible for merging, it:
  - MUST address all feedback provided during the review process.
  - MUST be approved by all team leads (of Status Desktop and Mobile).
  - MUST be approved by all members of the @status-im/status-go-guild
  GitHub team.
  - MUST receive a minimum of six approvals from CCs.

# Policy Overrides

> On rare occasions, circumstances may necessitate that an established

Reviewer 5: Missing newline

# Purpose

Policy Zero establishes the foundational guidelines for creating,
reviewing, and maintaining policies in the `status-go` Git repository.
This policy aims to create a collaborative, inclusive, and transparent
process for defining repository policies, specifically regarding how
developers engage with and contribute to the `status-go` repository.

# Submitting a Policy Proposal

- Any individual MAY propose a new policy.
- Policy ideas SHOULD be discussed with contributors who wish to have
a voice in how the repository operates, including Core Contributors
(CCs) and external contributors.
- All policies MUST be submitted to the `_docs/policies`
 directory as a pull request (PR) within the `status-go` repository.
- All policies MUST be in Markdown format.
- Policy file names MUST follow [File name conventions for ADRs](https://github.com/joelparkerhenderson/architecture-decision-record?tab=readme-ov-file#file-name-conventions-for-adrs), e.g. `submitting-policy.md`.

# Review and Approval Process

> The core function of the review and approval process for policy

Reviewer 4: I feel that the descriptions for each point, while very well written, may add a bit of noise, as they overlap with the README.md or are already clear from the title itself. For this particular point, it seems the 'Collective Agreement' section already covers it well.

Author: Ok that is fair. How about for this section I replace the description with a new item:

```
- Reviewers SHOULD keep in mind the principles communicated in 
[the README.md](README.md)
```

Reviewer 4: Hm, perhaps we could include a mention of README.md under the Purpose section? For example: "For a more detailed description, please refer to [the README.md](README.md)." This would then implicitly apply to all points.

must aim to achieve broad contributor support and give key stakeholders
a chance to gain context of the policy requirements.

- Policy PRs SHOULD be reviewed by as many contributors as possible
who wish to engage in the process.
- Any CC MAY review, approve, and/or request changes to a policy
proposal PR.
- For any policy PR to be eligible for merging, it:
  - MUST address all feedback provided during the review process.
  - MUST be approved by all team leads (of Status Desktop and Mobile).
  - MUST be approved by all members of the @status-im/status-go-guild
  GitHub team.
  - MUST receive a minimum of six approvals from CCs.

# Policy Overrides

On rare occasions, circumstances may necessitate that an established
policy is circumvented when merging a feature PR. This is considered
an **override** and MUST follow the process outlined below to ensure
transparency and collective agreement:

- Any override MUST be documented in textual form in the circumventing
feature PR and MUST include:
  - The specific policy being overridden,
  - The rationale for taking this action,
  - The potential risks and impacts of the override, and
  - Steps taken to minimise those risks.

> - Before proceeding, the override MUST be approved in writing in the

Reviewer 4: I am wondering whether "in writing" is really needed? Wouldn't PR approval be enough?

Author: Explicitly acknowledging that you approve the PR to breaking one or more policy items creates a document trail that records that the override approvers were not just approving the PR, they were also specifically approving the override.

I'm imagining a case where someone approves a PR that is circumventing a policy, the approval itself doesn't communicate that the approver knew they were also approving the override.

I am overthinking this? Am I missing something?

Reviewer 4: That's good point, you are right. The more explicit the approval of exception the better.

> circumventing feature PR by:

Reviewer 4: I believe there might be some policies unrelated to PRs :thinking: A very abstract example could be something like a 'GitHub teams management policy' - not the best example, but I hope you get my point.

Author: This is a good point, what about replace "feature PR" with "non-policy PR"? I'm trying to communicate that the overriding PR is not another policy PR.

```suggestion
circumventing non-policy PR by:
```

Let me know if I am not getting your point.

Author: Perhaps I should add a new item?

```
- A policy MUST not override an other policy
```

Reviewer 4: What I meant is that policy can be in theory violated not necessarily by code (aka no PR triggered). I don't have many examples in my mind, but would be great to keep policy 0 generic enough to allow such cases.

  - At least one team lead from the Status Desktop or Mobile teams, AND

>   - At least one member of the @status-im/status-go-guild GitHub team.

Reviewer 4: Would it make sense to follow the form from "Review and approval process" for consistency?

```
- Before proceeding, the override:
  - MUST be approved by at least one team lead (of Status Desktop and Mobile).
  - MUST be approved by at least one member of the @status-im/status-go-guild
```

Author: Ok, that is reasonable. What do you think about the following?

```
- Before approving the non-policy PR, the override:
  - MUST be approved in writing by at least one team lead from the Status Desktop or Mobile teams, AND
  - MUST be approved in writing by at least one member of the @status-im/status-go-guild.
```

Reviewer 4: It looks good. Since the policy might not be related to PRs, I would remove the reference to 'non-policy PR.' This approach would cover both PR and non-PR changes, making the policy 0 more flexible.

- In exceptional circumstances if an override MUST be executed
immediately due to urgency, the action SHOULD be documented as soon

> as possible, and retrospective approval MUST be sought and recorded

Reviewer 4: Can you please elaborate what do you mean by retrospective approval?

Author: I'm beginning to feel doubt about this point. I wanted to make a provision to allow for flexibility in the case of a truly exceptional case, something where we needed to act immediately and justify the actions after. That is what I mean by retrospective approval, you do the work and merge it quick but after you receive approval and explain what happened and why.

I wonder if this would open the door to abuse though. What do you think? (or anyone else)

Reviewer 4: I believe approval from one team lead and one member of status-go-guild covers the exceptional cases to be honest - meaning no retrospective approval.

at the earliest possible opportunity.
- Policies MAY define additional rules for handling overrides, provided
these baseline requirements are also met.

# Policy Amendments and Archival

Policies can be amended or archived to ensure they remain relevant

> and aligned with contributor needs.

Reviewer 4: :no_bell: :pray:

Author: I've removed this
